### DISCUSSION TITLE
The `-externalip` command does not work with .onion addresses

### DISCUSSION THREAD

**Discussion by harshil21** [dogecoin/dogecoin#2496] [Q&A] [answered]
Hello there. 

I am trying to run a dogecoin full node with incoming connections. Unfortunately port forwarding/UPnP does not work on my router (or maybe my ISP has blocked it). As a result I want to use tor to run the node and accept incoming connections. I followed the guide setup at https://github.com/dogecoin/dogecoin/blob/master/doc/tor.md. I understand that I need to launch `dogecoind` with the arguments `-proxy`, `-externalip` and `-listen`. So I do:

`dogecoind -proxy=127.0.0.1:9051 -listen -externalip=mytoraddr.onion` (my tor port is set to 9051). 
However I get the error:
`Error: Cannot resolve -externalip address: 'mytoraddr.onion'`.

I can run the above without `-externalip` and it even works when I set my actual ipv4 address in that field. But then I don't receive any inbound connections. 

Also my tor address is definitely valid, since I can run a webserver and use that address to access it from another device (using 4G).

Relevant part of torrc file:
![relevant torrc](https://user-images.githubusercontent.com/37377066/131195393-aabfba2b-d8de-4504-b808-c3ba7c9ad0ec.png)

Output of `dogecoin-cli getinfo`:

`
{
  "version": 1140400,
  "protocolversion": 70015,
  "walletversion": 130000,
  "balance": [omitted],
  "blocks": 3872609,
  "timeoffset": -1,
  "connections": 7,
  "proxy": "127.0.0.1:9051",
  "difficulty": 3211216.531227365,
  "testnet": false,
  "keypoololdest": 1630091191,
  "keypoolsize": 97,
  "unlocked_until": 0,
  "paytxfee": 1.00000000,
  "relayfee": 0.00100000,
  "errors": ""
}
`
Tor Browser version: 10.5.5

Edit: I tried facebook's .onion address (facebookcorewwwi.onion) and it worked. Why isn't it working with mine?

**Comment by harshil21** [comment, marked as answer]
Okay I just figured it out. For some reason supplying an external Tor V3 address is not compatible with dogecoin core. I had to edit my `torrc` to use a V2 address (`HiddenServiceVersion 2`). This is surprising cause V3 addresses were introduced quite long back.

However, after all this, I'm still not receiving any inbound connections. Why is this so?

**Reply by patricklodder** [reply]
> However, after all this, I'm still not receiving any inbound connections. Why is this so?

Most nodes don't connect over tor, so it will take a very long time for tor nodes to find you, if any (maybe there is no one else using tor, especially since as you noted, the service addressing is not working properly.) 

This is interesting though, so I will spend some time and set up a couple of my relay nodes to dual-connect with tor so that there are some more nodes on the onion network. I'll ping you back once I have something set up.

**Reply by harshil21** [reply]
Hi, is there any update on this? Is there a way I can manually test if my dogecoin node is actually accepting incoming connections through tor? Is there a command or something for this?

**Reply by patricklodder** [reply]
Hi! Thanks for reminding me. I have an onion node up at `vd5vtdiicazsowlj.onion` - you should be able to connect to that with `dogecoin-cli addnode vd5vtdiicazsowlj.onion add` and see that connection within the results of `dogecoin-cli getpeerinfo`

I tested this with a `-onlynet=onion` node and that worked.

**Reply by harshil21** [reply]
> Hi! Thanks for reminding me. I have an onion node up at `vd5vtdiicazsowlj.onion` - you should be able to connect to that with `dogecoin-cli addnode vd5vtdiicazsowlj.onion add` and see that connection within the results of `dogecoin-cli getpeerinfo`

It worked! I can connect to your 1.14.5 node. 

> I tested this with a `-onlynet=onion` node and that worked.

Did you get any inbound tor connections other than mine (`jxoya4f3zjbacd4w.onion`)? I'd expect there to be many tor nodes for dogecoin, since for bitcoin, [bitnodes](https://bitnodes.io/nodes) shows that there's around ~7k tor nodes.

I only got one outbound connection (yours) with `-onlynet=onion`.
When I look at my `debug.log` I see many: `2021-10-15 06:21:17 Socks5() connect to seed.multidoge.org:22556 failed: host unreachable`.
it's the same as #1693, which blamed it on the tor exit nodes. Maybe this is why most tor nodes can't connect to me? I can't imagine that this is the case though

**Reply by patricklodder** [reply]
> It worked! I can connect to your 1.14.5 node.

Yay! 🥳

> Did you get any inbound tor connections other than mine?

I also have a connection with `gxz6anzts46v7cza.onion` - and that's it. I fear there's not many more nodes on tor for Dogecoin, we're not the dark web's favorite coin (and I see no reason why we should be.) So for Dogecoin, I feel that running a tor node is mostly to provide a means against censorship for other shibes.

> 2021-10-15 06:21:17 Socks5() connect to seed.multidoge.org:22556 failed: host unreachable.

Hmm that is odd. I don't see it but I use the built-in torcontrol routine. You followed the manual tor service setup, not the one that uses torcontrol, right? Because that seed address is the DNS seed on port 53, not a listening node with an open port 22556. Not sure how a DNS seed would return an onion address either - could only be in AAAA records? Did you configure that node as a seed node?

I'll look at this further, see if I can get some more insights into how we can do some tor bootstrapping.

**Reply by harshil21** [reply]
> Hmm that is odd. I don't see it but I use the built-in torcontrol routine. You followed the manual tor service setup, not the one that uses torcontrol, right? 

yep, setted up manually.

> Because that seed address is the DNS seed on port 53, not a listening node with an open port 22556. Not sure how a DNS seed would return an onion address either - could only be in AAAA records? Did you configure that node as a seed node?

Nope I didn't change anything related to seeding, just added your node using `addnode` and followed the proxy setup guide above. Right now I'm actually doing the IBD of Bitcoin Core and am also going to setup tor on it and see if the behaviour is better than Dogecoin's.

**Reply by harshil21** [reply]
so an update: 

Now running a full bitcoin node on Tor, I finally get 10 out/(2-6) In connections. The new `-addrinfo` command also tells me there's ~6k Tor nodes I can connect to. So I guess for now, I can close this saying that there's simply not enough Tor nodes in the Doge network.
